[Leong Ming Jian Eugene] iP

eugenelmj opened the pull request:
**EugeneDuke**

> It takes as much energy to wish as it does to plan. - Eleanor Roosevelt [(source)](https://www.brainyquote.com/quotes/eleanor_roosevelt_379411)

EugeneDuke gives you the liberty to organise your life, just the way you like! It is:

- text-based
- read your day tasks with a single click
- ~~EASY~~ SUPER EASY to customise 💯 
- the best thing? **FREE**, at **NO COST** 🤑 

All you need to do is:

1. download it from [here](https://github.com/nus-cs2103-AY2223S1/ip/pull/209).
2. insert your to-dos
3. let it work its magic and manage your tasks for you 😆 

Its key features are:

- [ ] Managing to-dos, events, and deadlines
- [ ] Reminders (to be incorporated soon!)
- [ ] Snooze Reminders (coming soon!)

If you're an aspiring programmer, you may tweak it to your liking and make it fully customised for yourself! 😉 

Here's a snippet of what you may see:

```
public static void main(String[] args) {
        try {
            new Duke("./src/duke.txt").run();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
}

```

Do leave a good review if you like EugeneDuke. Thank you! 😄

Commit:
Add Gradle support

eugenelmj committed:
Level-1

Initial draft

eugenelmj committed:
Level-2

Included Add, List

eugenelmj committed:
Level-3

Added mark, unmark, and a Task class

eugenelmj committed:
Level-4

Updated with ToDo, Deadline, and Event class

eugenelmj committed:
A-TextUiTesting

I tried running runtest.sh locally, and even though both the expected and output looked the same, it still failed the test. To come back to it again.

eugenelmj committed:
Level-5

Handled "todo" and "blah" errors

eugenelmj committed:
Level-6

Implemented Delete

eugenelmj committed:
Level-7

Implemented Save

eugenelmj committed:
branch-Level-8

Implemented Dates and Times

eugenelmj committed:
Merge branch 'branch-Level-8'

eugenelmj committed:
Test

eugenelmj committed:
Did MoreOOP (after much effort...)

eugenelmj committed:
A-Packages

eugenelmj committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

eugenelmj committed:
Did A-Gradle and A-JUnit

eugenelmj committed:
For A-JAR

eugenelmj committed:
A-JavaDoc

eugenelmj committed:
Level-9

eugenelmj committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Level-9

HakkaNgin reviewed (commented):
test

HakkaNgin commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,37 @@
+        throw new FileNotFoundException("File cannot be found!");
+    }
+
+}
```
Review 1st round: Adheres to coding standard

HakkaNgin commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public void mark(int idx) {
+        Task t = tasks.get(idx);
```
Can perhaps try to use more elaborate names such as "task" and "index" instead of "t" and "idx"

HakkaNgin commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+}
```
Looks good. Code follows coding standards

HakkaNgin commented on src/main/java/Duke.java:
```diff
@@ -1,173 +1,78 @@
-                line + "\n");
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
Follows coding standards

HakkaNgin commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,63 @@
+                String input = sc.nextLine();
+                String commandType = Parser.getCommandType(input);
+                if ((commandType.equals("MARK"))) {
+                    int markIdx = Integer.parseInt(input.substring(5)) - 1;
```
Perhaps a more readable variable name such as "indexToMark" would be better?

HakkaNgin commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,63 @@
+                    String eAction = input.substring(6, input.indexOf("/") - 1);
+                    ls.add(new Event(eAction, Parser.formatEventTime(input)));
+                } if (commandType.equals("DELETE")) {
+                    int deleteIdx = Integer.parseInt(input.substring(7)) - 1;
```
Similarly, you may want to consider "indexToDelete"

HakkaNgin commented on src/main/java/Task.java:
```diff
@@ -9,11 +9,11 @@ public Task(String action) {
     }
 
-    public void mark() {
+    public void markIsDone() {
```
Maybe "markAsDone" and "markAsNotDone" sounds clearer?

HakkaNgin commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,63 @@
+public class TaskList {
+
+    private ArrayList<Task> tasks;
+    private static final String line = "-----------------------------";
```
Might consider moving this to the Ui class and making it a static field

HakkaNgin commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,37 @@
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+
+public class Ui {
```
I think you might want to make some of the fields (esp the messages/ prompts) static final, since you won't need to change them and can access them directly via Ui.saysHi() etc.

HakkaNgin commented on src/duke/Duke.java:
```diff
@@ -1,4 +1,4 @@
-import main.java.*;
+package duke;
```
You might want to consider more packages, such as a util package for the "tool" classes (e.g. Ui, Storage, Parser) and another package for Tasks (Deadline, ToDo, Task etc.)

HakkaNgin commented on src/main/java/duke/Deadline.java:
```diff
@@ -12,16 +19,28 @@ public Deadline(String action, LocalDateTime by) {
+     * Gives the date of the deadline as output
+     * @return date (in dd-MMM-yyyy)
+     */
     public String getDate() {
```
Good implementation to allocate two different methods to process date and time

HakkaNgin commented on src/main/java/duke/Duke.java:
```diff
@@ -3,12 +3,24 @@
+/**
+ * The self-named main class that controls our lovely chatbot, Duke
+ *
+ * @author eugeneleong
```
Good practice to note down your name and version number in the javadocs

HakkaNgin commented on src/main/java/duke/DukeException.java:
```diff
@@ -1,5 +1,13 @@
 
+/**
+ * A self-made Exception class that is supposed to handle Exceptions
+ * Not used
```
Might want to consider using this exception to handle cases where the user input is invalid (i.e. is not a command/ inputs a task in the wrong format/ is some gibberish) -- with a custom string message

HakkaNgin commented on src/main/java/duke/Event.java:
```diff
@@ -12,15 +19,28 @@ public Event(String action, LocalDateTime at) {
+     * Gives the date of the event as output
+     * @return date (in dd-MMM-yyyy)
+     */
     public String getDate() {
```
seems that the "getDate" and "getTime" methods are repeated in Deadline.java and Event.java. Might want to consider shifting them to the Task.java class to avoid repetition, or create a new class to handle date and time.

HakkaNgin commented on src/main/java/duke/Storage.java:
```diff
@@ -17,7 +24,11 @@ public Storage(String filePath) {
+     * @return ArrayList<Task></Task>
+     * @throws FileNotFoundException if file cannot be found
+     */
     public ArrayList<Task> load() throws FileNotFoundException {
```
Name sounds slightly ambiguous. Maybe loadTasks() sounds clearer

HakkaNgin commented on src/main/java/duke/Storage.java:
```diff
@@ -51,7 +62,12 @@ public ArrayList<Task> load() throws FileNotFoundException {
+     * @param text to save
+     * @throws IOException if the user's input cannot be saved
+     */
+    public void save(String text) throws IOException {
```
Renaming it to saveToFile() may add greater clarity

HakkaNgin commented on src/main/java/duke/Ui.java:
```diff
@@ -11,25 +18,37 @@ public class Ui {
+    /**
+     * Our lovely Duke gives a sweet intro
+     */
     public void saysHi() {
```
Might be better if the name of the method is written in imperative form (sayHi instead of saysHi, sayBye instead of saysBye)

eugenelmj replied:
Indeed, I should improve on that (as well as the other comments you've made). Thank you for taking the time to go through this in detail and giving the good nuggets of feedback! Cheers!

HakkaNgin commented on src/main/java/duke/TaskList.java:
```diff
@@ -25,6 +25,21 @@ public void listAll() {
         System.out.println(line + "\n");
     }
 
+    public void find(String keyword) {
```
findTask() may add greater clarity

HakkaNgin reviewed (commented):
Generally looks good! Keep up the work :)

junwei-tan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+    private Ui ui;
+
+    /**
+     * Loads Duke up with the appropriate list of tasks
```
Perhaps a full stop could be added at the end of each line as per coding standard.

junwei-tan commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     * @return date (in dd-MMM-yyyy)
+     */
+    public String getDate() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MMM-yyyy");
```
I appreciate your use of our local date-time format rather than the default American format.

junwei-tan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,59 @@
+     * @param input by the user
+     * @return type of command
+     */
+    public static String getCommandType(String input) {
```
Perhaps a switch block will make the code more succinct.

junwei-tan reviewed (commented):
Overall great job!! I like that your JavaDocs are succinct.

eugenelmj committed:
A-CheckStyle

Updates made with peer comments and CheckStyle

eugenelmj committed:
Starting Level-10

eugenelmj committed:
Skeleton for GUI

Made changes to original code and commented out parts of code in Duke.java.

eugenelmj committed:
Added Level-10 GUI

eugenelmj committed:
Rename mainClassName in build.gradle for shadowJar

eugenelmj committed:
Add A-CodeQuality

Review code quality to adhere to coding standard

There are several parts which were not tidied up; in particular, the "//Fallthrough" comment was missing for the 'switch' statement, where I intended for it to fall through, but did not write an explicit "Fallthrough" comment. I will continue to make sure the subsequent codes closely adhere to the accorded coding standard.

eugenelmj committed:
More to add to A-CodeQuality

Updated CodeQuality based on comments left on code

eugenelmj committed:
Add A-Assertions

Changes have been made to assert that the various parts of the 'process' command in Parser class is giving the correct output

eugenelmj committed:
Merge pull request #1 from eugenelmj/branch-A-Assertions

Add A-Assertions

eugenelmj committed:
Update Parser.java

I accidentally wrote the assertions on this branch instead, and I have since removed them.

eugenelmj committed:
Merge branch 'master' into branch-A-CodeQuality

eugenelmj committed:
Merge pull request #2 from eugenelmj/branch-A-CodeQuality

Branch-A-CodeQuality

eugenelmj committed:
Update A-CodeQuality

eugenelmj committed:
Merge pull request #3 from eugenelmj/branch-A-CodeQuality

Update A-CodeQuality

eugenelmj committed:
Update A-Assertions

eugenelmj committed:
Merge branch 'master' into branch-A-Assertions

eugenelmj committed:
Merge pull request #4 from eugenelmj/branch-A-Assertions

Update A-Assertions

eugenelmj committed:
Add C-Tagging

Make our chatbot Duke more fun to use!

There are many ways to make the app more engaging, and what better way to do so than to allow #tagging! This current version enables the user to add tags to their tasks (e.g. project meeting #yay!). Users can add as many tags as they like, and express their personal feelings via the tags. However, users are not able to remove tags - stay tuned for that in a future update! :smile:

eugenelmj committed:
Added SayHi() functionality

eugenelmj committed:
Enhance UI for A-BetterGui

Improve overall readability

Adjustments have been made to make the whole list of tasks appear, and not be limited by the text window. Colours and fonts have been adjusted to keep chatbot simple and easy to read.

eugenelmj committed:
Add Personality for A-Personality

Slight dialog tweaks have been made to make the PoliteDuke bot seem more friendly. Also, cleaned up dead code to improve code readability. Can do more tweaks in the future.

eugenelmj committed:
Adjustment to Event Class

Corrected the label as "E" instead of "D" (for Deadline)

eugenelmj committed:
Add Ui.png

eugenelmj committed:
Add A-UserGuide

eugenelmj committed:
Update Ui.png

eugenelmj committed:
Touch-up to Storage.java
